### PR TITLE
Fix kubectl run --dry-run with -i or --attach

### LINKED ISSUE #30081: kubectl run --dry-run cannot be used when -i is specified
While trying to figure out why a `--overrides` parameter was not parsed as I would have expected it to, I just threw in a `--dry-run` but saw the following error message:

```
$ kubectl run -i --generator=run-pod/v1 bug --image=nginx --dry-run 
error: an empty namespace may not be set when a resource name is provided
```

Turns out that one has to remove the `-i` flag in order to use `--dry-run`:

```
$ kubectl run --generator=run-pod/v1 bug --image=nginx --dry-run    
pod "bug" created
```

I’m wondering:
1. Why that is, i.e. why is `-i` detrimental to `--dry-run`? That seems unintuitive.
2. Whether the error message can be improved. Namespaces really don’t seem to have anything to do with what I’m seeing here.

Thanks!

### PR DESCRIPTION
**Pull request by ivan4th**
Don't attempt to attach to the pod in dry run mode.
Fixes #30081

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/34152)

### CODE REVIEW

**Comment by ivan4th** [comment]
ping...

**Comment by janetkuo** [comment]
I'd suggest moving this part from `generateService` to `createGeneratedObject`:
```go
	if cmdutil.GetFlagString(cmd, "output") != "" || cmdutil.GetDryRunFlag(cmd) {
		return f.PrintObject(cmd, mapper, obj, out)
	}
	cmdutil.PrintSuccess(mapper, false, out, mapping.Resource, args[0], cmdutil.GetDryRunFlag(cmd), "created")
```
and then we no longer need this in `Run`. We can then just return on dry-run right after expose. Please also add a test for this `-i --dry-run` case (in hack/make-rules/test-cmd.sh for example).

If we can abstract the `attach` part to its own function that'd also be great (optional).

**Comment by ivan4th** [comment]
@janetkuo made these changes. As of extracting `attach` part, I think something along the lines of `ExecOptions` in [exec.go](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/exec.go) needs to be done, i.e. extracting flags etc. into a struct to make the code cleaner. Multiple functions pulling cmd flags don't look good IMO. I can fix this in a followup PR.

File: pkg/kubectl/cmd/run.go [hunk 1]
```diff
@@ -570,20 +574,11 @@ func generateService(f cmdutil.Factory, cmd *cobra.Command, args []string, servi
...
-	}
-	cmdutil.PrintSuccess(mapper, false, out, mapping.Resource, args[0], cmdutil.GetDryRunFlag(cmd), "created")
-
-	return nil
+	_, _, _, _, err := createGeneratedObject(f, cmd, args, generator, names, params, cmdutil.GetFlagString(cmd, "service-overrides"), namespace, out)
+	return err
 }
 
-func createGeneratedObject(f cmdutil.Factory, cmd *cobra.Command, generator kubectl.Generator, names []kubectl.GeneratorParam, params map[string]interface{}, overrides, namespace string) (runtime.Object, string, meta.RESTMapper, *meta.RESTMapping, error) {
+func createGeneratedObject(f cmdutil.Factory, cmd *cobra.Command, args []string, generator kubectl.Generator, names []kubectl.GeneratorParam, params map[string]interface{}, overrides, namespace string, out io.Writer) (runtime.Object, string, meta.RESTMapper, *meta.RESTMapping, error) {
```
Comment (janetkuo) [inline pkg/kubectl/cmd/run.go]: Just pass resource name (`string`) instead of passing `args []string`

File: pkg/kubectl/cmd/run.go [hunk 2]
```diff
@@ -649,5 +646,15 @@ func createGeneratedObject(f cmdutil.Factory, cmd *cobra.Command, generator kube
...
 		}
 	}
+
+	if cmdutil.GetFlagString(cmd, "output") != "" || dryRun {
+		err = f.PrintObject(cmd, mapper, obj, out)
+		if err != nil {
+			return nil, "", nil, nil, err
+		}
+	} else {
+		cmdutil.PrintSuccess(mapper, false, out, mapping.Resource, args[0], cmdutil.GetDryRunFlag(cmd), "created")
```
Comment (janetkuo) [inline pkg/kubectl/cmd/run.go]: just use `dryRun`

**Review by janetkuo — CHANGES_REQUESTED** [review]

**Comment by janetkuo** [comment]
We print a redundant message now. Just need to remove the print message from the bottom of `Run`

```console
$ kubectl run nginx --image=nginx 
deployment "nginx" created
deployment "nginx" created
```

**Comment by ivan4th** [comment]
Fixed. Also, I'll try to make it all prettier in a followup PR.

**Comment by ivan4th** [comment]
#35064
@k8s-bot unit test this

**Comment by ivan4th** [comment]
Will retry e2e a bit later

**Comment by ivan4th** [comment]
@k8s-bot unit test this

**Comment by ivan4th** [comment]
@k8s-bot gci gke e2e test this

**Review by janetkuo — APPROVED** [review]

**Comment by ivan4th** [comment]
@k8s-bot gci gke e2e test this

**Comment by ivan4th** [comment]
@k8s-bot node e2e test this

**Comment by ivan4th** [comment]
@k8s-bot node e2e test this

**Comment by ivan4th** [comment]
Failed to merge due to CI infra problems :( #35732 is merged now, it makes `kubectl run` fail with these flag combinations. I think that adding `--dry-run` ignoring `--attach` / `-i` is more intuitive, e.g. when I add `--dry-run -o yaml` to a long `kubectl run` command I typed before I expect it to display objects it generates without making me remove some of the other flags. Still, I expect such discussion to be difficult so I'm inclined to give up on this PR.

**Comment by janetkuo** [comment]
> I'm inclined to give up on this PR.

Ah, agree. Since #35732 already fixed #30081, we can close this one.

**Pull request closed without merging**